### PR TITLE
Basic work for right sidebar

### PR DESCRIPTION
**Pull request by PVince81**
Adds right sidebar with registrable panels (still WIP)

For https://github.com/owncloud/core/issues/17655

~~Note: file actions are currently disabled for demo purposes. When clicking on a file it will show the sidebar but not perform any action.~~
- [x] TODO: rewire click events to have the default file action work on the left side of the row but not the right side
- [x] TODO: implement tabs logic
- [x] TODO: implement default tab
- [x] TODO: styling
- [x] TODO: sidebar must close when unselecting file or selecting multiple files ?

### CODE REVIEW

**Comment by PVince81** [comment]
Basically this brings the following:
- `DetailsView` which is the right sidebar / panel. On this class it's possible to register subviews
- `DetailFileInfoView` is the base class for any file info block to be displayed in the top panel. Needs to be registered with `detailView.addDetail(...)`
- `MainFileInfoDetailView` implements `DetailFileInfoView` to provide the basic info like thumbnail, file name
- `DetailTabView` is the base class for info tabs. Needs to be registered with `detailView.addTab(...)` (the logic isn't implemented yet)
- styles can be tweaked in `detailsView.css`

I know... a lot of classes and not so fancy names...

**Comment by PVince81** [comment]
@DeepDiver1975 @jancborchardt @blizzz @oparoz @rullzer

**Comment by PVince81** [comment]
@jancborchardt I think it's still too early to tweak the styles. I'll let you know when it's ready :smile:

File: apps/files/js/mainfileinfodetailview.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+			if (!this._template) {
+				this._template = Handlebars.compile(TEMPLATE);
+			}
+
+			if (this._fileInfo) {
+				this.$el.append(this._template(this._fileInfo));
+				var $iconDiv = this.$el.find('.thumbnail');
+				// FIXME: use proper way, this is only for demo purposes
+				FileList.lazyLoadPreview({
```
Comment (oparoz) [inline apps/files/js/mainfileinfodetailview.js]: This won't work as there is a hardcoded value of 36x36 as a fallback. We keep running into this issue.
Besides, I don't think we need lazy loading here. Just point to the preview using `generatePreviewUrl`and the browser will show it when ready.

Comment (oparoz) [inline apps/files/js/mainfileinfodetailview.js]: The mockup indicates 50x50, but that's too small for the prevew of a selected file. Maybe we should load a 200x200 preview and let the browser resize it based on the available width, if the sidebar is responsive.

Comment (PVince81) [inline apps/files/js/mainfileinfodetailview.js]: I'll add it as todo. The FIXME already says that what I did here is wrong :wink:

**Comment by PVince81** [comment]
- [ ] TODO: load previews with the correct API, maybe with 200x200 format and reduce it to 50x50 ? ( @oparoz feel free to comment/rectify)

File: apps/files/js/mainfileinfodetailview.js (hunk 1 above)
Comment (oparoz) [inline apps/files/js/mainfileinfodetailview.js]: Ah, yeah, I only looked at the todos :D

**Comment by PVince81** [comment]
I realize that the file info details (the top view) cannot be extended block-wise.
Maybe the easiest is to simply pass the jQuery element to plugins so plugins can embed themselves wherever. Here you can see that "Owner: " is in its own block. That one is provided by the "files_sharing" app.

I'm not too happy yet about the way how to provide tab view and detail view, it seems quite cumbersome to implement/extend a big class just for a few lines.

Also due to the problematic loading order of JS files I had to do this: https://github.com/owncloud/core/pull/17656/files#diff-593f07f9e5d32f4d3fdebbfb2fa5b71aR144

**Comment by PVince81** [comment]
- [ ] make it possible to hide tabs based on content type ? (ex: folders)

**Comment by PVince81** [comment]
- [ ] hightlight current row
- [ ] full height scrollbar on the sidebar (let the contents flow, no inner scrollbars)

**Comment by PVince81** [comment]
- [ ] BUG: hiding animation leaves horizontal scrollbar
- [ ] TODO: make favorite action work

**Comment by PVince81** [comment]
- [ ] TODO: unit tests for the panels logic

**Comment by PVince81** [comment]
I'm not sure if we want to do the share dialog task for now.
Maybe we can first get this in a mergable state and the sidebar will initially have no tabs shown at all.
So we need to get the interaction with the file list right.

When testing, please make sure to test the sidebar in the other views: "Shared with you", "Deleted files", etc. :smile:

**Comment by blizzz** [comment]
> Basically this brings the following:
- `DetailsView` which is the right sidebar / panel. On this class it's possible to register subviews
- `DetailFileInfoView` is the base class for any file info block to be displayed in the top panel. Needs to be registered with `detailView.addDetail(...)`
- `MainFileInfoDetailView` implements `DetailFileInfoView` to provide the basic info like thumbnail, file name
- `DetailTabView` is the base class for info tabs. Needs to be registered with `detailView.addTab(...)` (the logic isn't implemented yet)
- styles can be tweaked in `detailsView.css`

> I know... a lot of classes and not so fancy names...

Nothing against many classes, if they make the whole feature better readable, understandable and maintainable. Of course this can be done, but I do not see it here.

The naming however should make you roughly understand what this is it about. DetailsView is pretty general, more obvious would be to have Sidebar or Panel in it's name indeed, but would cost you flexibility to change it in the future to a e.g. bumping circle. As it shows details, I am good with the naming.

Most importantly, the structure makes sense to me. All question that occurred to me were well answered in this approach.

**Comment by PVince81** [comment]
@blizzz thanks for the feedback :smile: 
I'd say let's keep this approach then and we can adjust later if needed.

**Comment by PVince81** [comment]
- [ ] fix scrolling: the top of the sidebar must always stay visible when scrolling down the file list (ex: when getting details of a file on page 12). Add a second scrollbar inside the sidebar to scroll its whole content.

**Comment by PVince81** [comment]
right-sidebar

**Comment by PVince81** [comment]
- [ ] TEST: public page

**Comment by PVince81** [comment]
I changed the scrolling behavior: now the sidebar has its own scrollbar.
Unfortunately it doesn't look very good when there are two scrollbars, but there might not be any better way around, unless we add an inner scrollbar inside the tab content instead.

@jancborchardt I think you can start tweaking the styles now and test the interaction :smile: 
I left the debug tabs so you can test the scrolling.

right-sidebar2

**Comment by PVince81** [comment]
- [ ] TEST in internet explorer

**Comment by PVince81** [comment]
- [ ] BUG: fix display name in trash bin (use fileInfo.displayName)

**Comment by jancborchardt** [comment]
@PVince81 please _do not_ use jQuery UI here, as we want to ideally get rid of it. Simply use buttons, and I will style them properly similar to the database chooser on installation.

**Comment by PVince81** [comment]
The reason I used tabs is just because I wanted to save time and saw a working example on the LDAP settings page.

- [ ] TODO: implement own tabs plugin/logic

**Comment by jancborchardt** [comment]
Ok. Just noting because we definitely want to do it like the db switcher on the install page.

**Comment by jancborchardt** [comment]
@PVince81 @owncloud/designers after chatting with @karlitschek we think it’s really better to have the right sidebar not only overlay the content but really move it away. Otherwise it’s only modal, actions and info below it is hidden.

We also need to pay attention to other apps which use similar things. Calendar+ by @libasys uses a right sidebar for Tasks integration. Same for Tasks by @raimund-schluessler & @Henni – there the right sidebar is for details of a task.
We should define the style and interaction for this sidebar in core so that apps can adapt it.

**Comment by raimund-schluessler** [comment]
Interesting work.

I would also vote for moving the content, so you can still access / see actions and info currently below the sidebar. This also solves the scroll bar problem since the scroll bar for the content would then be on the left of the sidebar (see tasks app as example).

**Comment by oparoz** [comment]
Bear in mind that some apps need to redraw the entire view when you change the viewport's width. 
If it penalises just Gallery, then maybe that app can work around the problem, but there may be others.

**Comment by jancborchardt** [comment]
@oparoz good point, Gallery isn’t fully responsive in this way. Probably also nothing really to do about it than trying to make the redraw faster? Or keep showing the old content while redrawing in the background and replacing it then?

**Comment by oparoz** [comment]
Well, Gallery is fully responsive. The problem I see is more a UX one with a redrawn photowall every time a user clicks on the share button. 
This is not really something that one does often, even if each picture can be shared, so maybe a non-issue? I'm just mentioning it in case this could happen to other apps as well.

**Comment by MorrisJobke** [comment]
```
16:54:12 PhantomJS 1.9.8 (Linux 0.0.0) OCA.Files.DetailsView tests file info detail view updates registered tabs when fileinfo is updated FAILED
16:54:12 	Expected 4 to equal 2.
16:54:12 	    at /var/jenkins/workspace/pull-request-analyser-ng-simple@12/label/SLAVE/apps/files/tests/js/detailsviewSpec.js:62
16:54:12 .
16:54:12 PhantomJS 1.9.8 (Linux 0.0.0) OCA.Files.DetailsView tests tabs updates registered tabs when fileinfo is updated FAILED
16:54:12 	Expected 4 to equal 2.
16:54:12 	    at /var/jenkins/workspace/pull-request-analyser-ng-simple@12/label/SLAVE/apps/files/tests/js/detailsviewSpec.js:91
16:54:16 PhantomJS 1.9.8 (Linux 0.0.0) OCA.Files.MainFileInfoDetailView tests rendering displays basic info FAILED
16:54:16 	Expected 'July 16, 2015 9:02 PM' to equal 'July 17, 2015 3:02 AM'.
16:54:16 	    at /var/jenkins/workspace/pull-request-analyser-ng-simple@12/label/SLAVE/apps/files/tests/js/mainfileinfodetailviewSpec.js:54
```

**Comment by Henni** [comment]
Awesome work!
Looking forward to adopt this in the tasks app.

**Comment by blizzz** [comment]
I'll continue while @PVince81 is off next week.

So, If I understand you correct, you would introduce styles for a new #app-sidebar css class. It will take the style of mostly what is .detailsView right now, with adjustments to how it should behave – not overlapping, but reducing the size of the app-content respectively it's neighbours. Unfortunately I would not know how to modify css to get this behaved properly. I'd be happy about pointers from design gurus.

**Comment by jancborchardt** [comment]
@blizzz you can set the #app-content width to 70% while setting the #app-sidebar width to 30%. We do it similarly in the Mail app with the message list and the detail view. Let me know when you need help.

**Comment by blizzz** [comment]
@jancborchardt code pointers would be nice. I tried to setup mail to see it in action, but cannot use it due to some Horde issues.

**Comment by jancborchardt** [comment]
@blizzz setting up the Mail app should be no problem if you just clone it and pull in the dependencies via composer [as said in the readme](https://github.com/owncloud/mail/#developer-setup-info).

Then inspect it, that makes more sense than code pointers. In any case: https://github.com/owncloud/mail/blob/master/css/mail.css#L222-L223 (30% width for message list) and https://github.com/owncloud/mail/blob/master/css/mail.css#L551-L559 (70% width for detail view.

For this of course it should be main content 70% width (that is, 70% of the #app-content), and 30% width for the right sidebar.

**Comment by blizzz** [comment]
@jancborchardt i read readme, and that all went good, but then i am stopped at https://github.com/owncloud/mail/issues/897

Thanks for the pointers. I was having fun with JS Karma tests today :unamused: I will get the design change done tomorrow :yum:

**Comment by blizzz** [comment]
Tests not working, because Jenkins, as far as i see.

**Comment by blizzz** [comment]
@jancborchardt ok, got  it so far. You may have a look? What's not good, yet, is it'll be too small on small screens, especially on mobile.  How was it supposed to behave in that case?

If the approach is ok, I'll transfer relevant parts to core apps css and js.

**Comment by jancborchardt** [comment]
@blizzz how do I invoke it? Sharing still shows a dropdown, as does versions.

**Comment by jancborchardt** [comment]
@blizzz regarding mobile screens, actually there we might need to have it overlap the content because of the limited width either way. Could be solvable by just CSS.

**Comment by blizzz** [comment]
> how do I invoke it? Sharing still shows a dropdown, as does versions.

@jancborchardt click on an empty space on the file row. Where still on foundations. Sharing and Version will move and be replaced in a later PR.

> regarding mobile screens, actually there we might need to have it overlap the content because of the limited width either way. Could be solvable by just CSS.

Roger.

**Comment by jancborchardt** [comment]
@blizzz some things:
* instead of having the class `detailsView` the sidebar should have the ID `app-sidebar` to fit with the rest. We do not use camelCase in classes or IDs.
* it should be over the controls bar as well, as in the prototype at https://github.com/owncloud/core/pull/16414#issue-77652540

**Comment by blizzz** [comment]
> it should be over the controls bar as well, as in the prototype at #16414 (comment) 

I tend to agree, still having some doubts as the breadcrumbs might become inaccessible in a deep directory (and the other controls as well), requires the user to click the close-x when he is done what he wanted to do here. Assuming this will not happen too often indeed, it's probably reasonable.

**Comment by blizzz** [comment]
* .detailsView is now #app-sidebar and resides in core/css/apps.css
* showing and hiding functions reside in core/js/apps.js
* overlaps in mobile view
* I added also min-width in standard view. It will overlap if the sidebar would be smaller than 300px.

* [ ] FIXME:  when mobile.css is effective, click on file row downloads file, instead of opening sidebar

**Comment by PVince81** [comment]
If we want the new sidebar to be usable by any app, we need to move it into core.
Also we then need to think about how to register sections.
Because any app could register a section for specific app.
So if for example the news app had the sidebar, and another app wants to register an extra panel, it should then call something like `registerPanel("news", panelInstance)`.

**Comment by blizzz** [comment]
We'll stick with having the basics in core and move styles and utility functions in a next iteration, after having collected experience and stabilized and polished what we're introducing here, so we can go with mature APIs later.

**Comment by PVince81** [comment]
Agreed with that.

**Comment by MorrisJobke** [comment]
* [ ] please don't use the jQuery widget thingy. It's super ugly and doesn't really fit into the WebUI.

**Comment by PVince81** [comment]
@blizzz nice work on making the sidebar insert itself under the controls and shrink the list width.

However there seems to be a new issue:
Before scroll:
rsidebar-top

After scroll:
rsidebar-top

@jancborchardt should the sidebar stay sticky together with the controls bar then ?
This means that moving the scrollbar only ever scrolls the contents of the file list.

**Comment by PVince81** [comment]
- [ ] BUG: sidebar scrolls with contents

**Comment by PVince81** [comment]
Goodbye jquery.ui, now with own tabs implementation.


Please note that I added some generic styles in apps.css, so the tabs styles will be reusable: https://github.com/owncloud/core/commit/ef4c95dd02e53936f668fa9196d6401ba8a0a307#diff-d28ce211f72cb6409a66e6c8ab52d723R596

**Comment by PVince81** [comment]
rsidebar-tabs

**Comment by jancborchardt** [comment]
Pushed some style fixes. The app-sidebar needs to be position: fixed. This leads to us not being able to properly size the app-content through CSS – we can only approach it, and do it via JS, similarly as we size the controls bar. (We can use the same function as for the controls bar sizing since we need to adjust that for the right sidebar anyway.)

**Comment by jancborchardt** [comment]
Also, the favorite star toggle should work. :) Also, clicking on the thumbnail or the filename should open the viewer for the respective file. For folders, I’m not sure if clicking the thumbnail should do anything.

**Comment by PVince81** [comment]
@jancborchardt I'd like to do the favorite star separately because it needs some deep JS magic (interaction between sidebar and file model) and I don't want to do it in a hacky way. I might remove it / hide it before we merge this.

**Comment by PVince81** [comment]
@jancborchardt clicking on a folder icon would open the folder then ? The thumbnail would use the default action as registered by apps.

**Comment by PVince81** [comment]
- [ ] thumbnail click must trigger default action

**Comment by oparoz** [comment]
> Also, clicking on the thumbnail or the filename should open the viewer for the respective file

Since the thumbnail offered is tiny, I would go with 2 different actions:
* One action to grow the thumbnail to 200x200, pushing the tabs down. This can be done with nice an overlay icon, like the one you see in video viewers when you want to make the viewport larger
* One action to open the file in it's default viewer. This can stay as the default action when clicking on the image

**Comment by jancborchardt** [comment]
> I might remove it / hide it before we merge this.

@PVince81 yeah, probably best call about the fav star.

> clicking on a folder icon would open the folder then ? The thumbnail would use the default action as registered by apps.

Yup.

> Since the thumbnail offered is tiny, I would go with 2 different actions: One action to grow the thumbnail to 200x200, pushing the tabs down.

@oparoz not sure we want to go with a mode like that. I do agree though that the 50px thumbnail is a bit small for that we use it in a detail view. How about increasing the size to 75px?

**Comment by PVince81** [comment]
Hmmm, making the thumbnail clickable has the same issue as for the favorite star.
The problem is that the "FileActions" object is old and broken and isn't portable. It isn't enough to inject the file action object into all components, one would require to inject the FileList object and also the $tr element and pass it along to each view inside the sidebar.
Which means that in total we need to inject:
- FileInfo object (already there)
- FileActions object
- FileList object
- $tr object

This is too much, I'll try and find a way to "package" this or refactor it to a better class design.
Ideally we should only need to inject a single object into the view. Then the view will be able to read the FileInfo from that object and also trigger actions.

**Comment by jancborchardt** [comment]
@PVince81 ok, making thumbnail interactive and duplicating the favorite icon there is not very important here because it’s duplication anyway – it’s mainly cherry on top.

What is important though for the sidebar is to put the sharing view into it, because without it it’s quite useless. ;) And then we need to continue with Versions, Activity & Comments.

**Comment by blizzz** [comment]
Sharing will be a seperate PR.

**Comment by PVince81** [comment]
Before the sharing or other views can be added in we need to merge this branch first.

**Comment by PVince81** [comment]
@owncloud-bot retest this please

**Comment by jancborchardt** [comment]
kk. @owncloud/designers please also give this a review! :)

**Comment by PVince81** [comment]
@jancborchardt the sidebar position and scrolling behavior is correct. However the main list's scrollbar isn't visible any more in Chromium. Did you intend the scrollbar to move on the blank space between the file list and scrollbar ? (might not be possible unless we change the scrollContainer for the list to not be the full page)

**Comment by jancborchardt** [comment]
> Did you intend the scrollbar to move on the blank space between the file list and scrollbar ?

@PVince81 yeah, that’s probably what we need to do. Also, as said above the width of the container needs to be computed by JS as a result of the positioning fix – there should not be any blank space between the file list and the right sidebar of course.

If making the scrollbar display there is not possible, we can also of course absolutely position the app-sidebar a bit from the right (using getScrollbarWidth). But that will look strange once the sidebar gets longer and thus a scrollbar – resulting in confusing double scrollbar on the right.

**Comment by PVince81** [comment]
I'd still vote for merge first as it brings the base classes needed for further dev and fix the layout afterwards as it might take some time.

- [ ] fix layout / scrolling

**Comment by PVince81** [comment]
@jancborchardt I found a non-intrusive fix that doesn't require revisiting the DOM layout.
See https://github.com/owncloud/core/commit/abe2213ba98c84ec03b877d2fa6a852f5203a291

Basically the sidebar width is set to 27% and the margin-right of app-content set to 27% too...

**Comment by PVince81** [comment]
Jenkins PR: https://github.com/owncloud/core/pull/18103

**Comment by PVince81** [comment]
Looks like these tests are too time-dependent:
```
Expected '幾秒前' to equal 'a few seconds ago'.
09:38:00 	    at /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files/tests/js/mainfileinfodetailviewSpec.js:56
09:38:00 	Expected '2015年7月16日晚上9點02分' to equal 'July 17, 2015 3:02 AM'.
09:38:00 	    at /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files/tests/js/mainfileinfodetailviewSpec.js:57
```

I'll have a look.

**Comment by PVince81** [comment]
Weird, looks like something put the test server in a different timezone and locale !? (taiwanese)

**Comment by PVince81** [comment]
@jancborchardt I also wonder whether the sidebar should also open when clicking on the other columns like mtime and size ?

**Comment by PVince81** [comment]
Ok now fighting against a timezone issue:
```
10:59:42 PhantomJS 1.9.8 (Linux 0.0.0) OCA.Files.MainFileInfoDetailView tests rendering displays basic info FAILED
10:59:42 	Expected 'July 16, 2015 9:02 PM' to equal 'July 17, 2015 3:02 AM'.
10:59:42 	    at /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files/tests/js/mainfileinfodetailviewSpec.js:57
```

**Comment by PVince81** [comment]
Hmmm, same problem...
```
12:35:39 PhantomJS 1.9.8 (Linux 0.0.0) OCA.Files.MainFileInfoDetailView tests rendering displays basic info FAILED
12:35:39 	Expected 'July 16, 2015 9:02 PM' to equal 'July 17, 2015 3:02 AM'.
12:35:39 	    at /var/jenkins/workspace/pull-request-analyser-ng-simple@4/label/SLAVE/apps/files/tests/js/mainfileinfodetailviewSpec.js:57
```

**Comment by blizzz** [comment]
My attempt to fix the test. Also rebased.

**Comment by PVince81** [comment]
JS unit tests fail:
```
07:24:11 	Expected '' to equal 'One.txt'.
07:24:11 	    at /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/apps/files/tests/js/mainfileinfodetailviewSpec.js:54
07:24:11 	Expected '' to equal '123456789 bytes'.
07:24:11 	    at /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/apps/files/tests/js/mainfileinfodetailviewSpec.js:56
07:24:11 	Expected '' to equal 'July 16, 2015 9:02 PM'.
07:24:11 	    at /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/apps/files/tests/js/mainfileinfodetailviewSpec.js:58
```

I'll look into this.

**Comment by PVince81** [comment]
I hereby approve all commits from @blizzz and @Henni and @jancborchardt with this :+1: 

Please review my commits too so we can get this merged.

**Comment by blizzz** [comment]
And so am I honoured to give a :+1: to all commits from @PVince81 @Henni and @jancborchardt :fireworks:

**Comment by PVince81** [comment]
Unless there is any objection, I will merge this as soon as Jenkins gives the green light. :bulb:

**Comment by PVince81** [comment]
Here we go.

Please raise tickets separately.

I'll soon make a ticket for the file actions (favorites + click on thumbnail)

**Comment by PVince81** [comment]
Raised ticket to implement the missing actions (thumbnail click + favorite click): https://github.com/owncloud/core/issues/18169

**Comment by jancborchardt** [comment]
@PVince81 great work man! :)

**Comment by DeepDiver1975** [comment]
@PVince81 branch can be deleted?

**Comment by PVince81** [comment]
>  branch can be deleted?

Yes. Strangely Github didn't show me the button yesterday.

**Comment by MorrisJobke** [comment]
> Yes. Strangely Github didn't show me the button yesterday.

On mobile? Because on mobile it isn't present

**Comment by PVince81** [comment]
No, on desktop web UI. It was probably a temporary glitch.

**Pull request merged**